README.md
## uxcore-user-guide

React user guide

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test Coverage][coveralls-image]][coveralls-url]
[![Dependency Status][dep-image]][dep-url]
[![devDependency Status][devdep-image]][devdep-url]
[![NPM downloads][downloads-image]][npm-url]

[![Sauce Test Status][sauce-image]][sauce-url]

[npm-image]: http://img.shields.io/npm/v/uxcore-user-guide.svg?style=flat-square
[npm-url]: http://npmjs.org/package/uxcore-user-guide
[travis-image]: https://img.shields.io/travis/uxcore/uxcore-user-guide.svg?style=flat-square
[travis-url]: https://travis-ci.org/uxcore/uxcore-user-guide
[coveralls-image]: https://img.shields.io/coveralls/uxcore/uxcore-user-guide.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/uxcore/uxcore-user-guide?branch=master
[dep-image]: http://img.shields.io/david/uxcore/uxcore-user-guide.svg?style=flat-square
[dep-url]: https://david-dm.org/uxcore/uxcore-user-guide
[devdep-image]: http://img.shields.io/david/dev/uxcore/uxcore-user-guide.svg?style=flat-square
[devdep-url]: https://david-dm.org/uxcore/uxcore-user-guide#info=devDependencies
[downloads-image]: https://img.shields.io/npm/dm/uxcore-user-guide.svg
[sauce-image]: https://saucelabs.com/browser-matrix/uxcore-user-guide.svg
[sauce-url]: https://saucelabs.com/u/uxcore-user-guide


### Development

```sh
git clone https://github.com/uxcore/uxcore-user-guide
cd uxcore-user-guide
npm install
npm run server
```

if you'd like to save your install time，you can use uxcore-tools globally.

```sh
npm install uxcore-tools -g
git clone https://github.com/uxcore/uxcore-user-guide
cd uxcore-user-guide
npm install
npm run dep
npm run start
```

### Test Case

```sh
npm run test
```

### Coverage

```sh
npm run coverage
```

## Demo

http://uxcore.alibaba.net/components/uxcore-user-guide?type=others

## Contribute

Yes please! See the [CONTRIBUTING](https://github.com/uxcore/uxcore/blob/master/CONTRIBUTING.md) for details.

## 使用方法

一个产品或者页面中可能有多个引导，需要给每个引导做一个key

```javascript
const blokingUserGuide = UserGuideFactory.getWithKey('1', {
  // 可以自定义参数
  isBlocking: true,
  assistType: 'SKIP',
  onComplete() {
    /** 引导结束，执行业务逻辑 **/
  },
});
```

```javascript
const UserGuide = UserGuideFactory.getWithKey('2', {
  // 可以自定义参数
  isBlocking: false,
  assistType: 'NO_REMIND',
  onAssistClick(step) {
    // 点击 了解更多 / 不再提醒 时的回调函数
    console.log(step);
    // 可以调用 stop 关闭当前引导。
    UserGuide.stop();
    /** 业务逻辑，例如存储不再提醒，下次不要让用户查看相关提醒了。 **/
  },
});
```
config 种含有的参数包括

| 参数名 | 类型 | 默认值 | 备注 |
| --- | --- | --- | --- |
| locale | string | `'zh-cn'` | 语言 |
| prefixCls | string | `'kuma-user-guide'` | class 前缀 |
| icon | string / React 组件 | undefined | 传递 string 时，以 uxcore-icon 的名字进行渲染；如果为 falsy 的值，则不会渲染；其他直接渲染，不包含外边框。 |
| className | string | `''` | 定制类名 |
| isBlocking | boolean | `true` | 是否阻塞UI|
| assistType | string | `undefined` | 辅助按钮 / 链接，可选值包括 `'SKIP'`: 跳过; `'LEARN_MORE'`: 了解更多; `'NO_REMIND'`: 不再提醒 |
| onAssistClick | function | `undefined` | 了解更多 / 不再提醒 点击时的回调函数 |
| onComplete | function | `undefined` | 结束时的回调 |


然后给这个引导添加步骤，每个步骤可以有4种选择
* 使用React HOC

```javascript
const Step1 = UserGuide.addUserGuide({
  dom: 'button',
  step: 1,
  icon: 'shanchu',
  hint: '我是第一步提示',
  type: 'ReactComponent',
});
```
* 使用DOM

```javascript
const Step1 = UserGuide.addUserGuide({
  dom: document.getElementById('app'),
  step: 2,
  hint: '我是第二步提示',
  type: 'HTMLElement',
});
```
* 使用一个函数返回DOM

```javascript
const Step1 = UserGuide.addUserGuide({
  getDom() { return document.getElementById('app'); },
  step: 3,
  hint: '我是第三步提示',
  type: 'HTMLElementMaker',
});
```
* 使用图片

```javascript
UserGuide.addUserGuide({
  step: 4,
  hint: '我是第四步的提示，我也没有对应的DOM',
  type: 'Image',
  top: 1800,
  left: 1000,
  image: 'https://img.alicdn.com/tfs/TB1TRNAllfH8KJjy1XbXXbLdXXa-240-240.png',
  imageHeight: 120,
  imageWidth: 120,
});
```

需要开始引导时，需要

```javascript
UserGuide.start();
```

start 可以传递参数:
| 参数名 | 类型 | 默认值 | 备注 |
| --- | --- | --- | --- |
| designMode | bool | false | 是否设计模式，如果是`true`，会展示所有的步骤，以供调试及预览使用。 |


需要关闭时，可以调用
```javascript
UserGuide.stop();
```

stop 可以传递参数
| 参数名 | 类型 | 默认值 | 备注 |
| --- | --- | --- | --- |
| callOnComplete | bool | true | 关闭时是否调用onComplete |

HISTORY.md
# history

## 0.2.1

* `FIXED` 多语言文案缺失
* `CHORE` 原包没有 npm 发布权限，从 user-guide 更名为 user-guide-v2

## 0.2.0
* `NEW`: `'SKIP'`, `'LEARN_MORE'`, `'NO_REMIND'` for `assistType`
* `NEW`: `isBlocking` for block or no-blocking user guide
* `NEW`: `icon` for icon in user guide step left
* `NEW`: callback `onAssistClick` when `'LEARN_MORE'`, `'NO_REMIND'` clicked

## 0.1.2
* `FIXED` #1 修正prefixCls, btn引用，less变量改名防止冲突

## 0.1.1

* `NEW` 初始化


tests/AitoUserGuide.spec.js
import expect from 'expect.js';
import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import UserGuide from '../src';

Enzyme.configure({ adapter: new Adapter() });

describe('UserGuide', () => {

});

src/UserGuide.js
/**
 * UserGuide Component for uxcore
 * @author buzhou
 *
 * Copyright 2015-2016, Uxcore Team, Alinw.
 * All rights reserved.
 */
const React = require('react');
const ReactDOM = require('react-dom');
const UserGuideStage = require('./UserGuideStage');
const scrollToTop = require('./scrollToTop');

const guideMap = {};

// HOC, 不是一个react component
export default class UserGuide {
  constructor(key, config) {
    this.key = key;
    this.locale = config && config.locale || 'zh-cn';
    this.prefixCls = config && config.prefixCls || 'kuma-user-guide';
    this.className = config && config.className || '';
    // 默认 true,
    this.isBlocking = config && config.isBlocking;
    if (this.isBlocking === undefined) {
      this.isBlocking = true;
    }
    this.assistType = config && config.assistType;
    this.onAssistClick = config && config.onAssistClick;
    this.onComplete = config && config.onComplete;
  }
  addUserGuide(guideProps) {
    const { step = 0 } = guideProps;
    if (!this.steps) {
      this.steps = [];
    }
    this.steps[step] = guideProps || {};
    // if (WrappedComponent instanceof HTMLElement) {
    //   this.steps[step].dom = WrappedComponent;
    //   return WrappedComponent;
    // }
    if (guideProps.type === 'ReactComponent') {
      return (props) => <guideProps.dom
        {...props}
        ref={(comp) => {
          this.steps[step].dom = comp;
        }}
      />;
    }
    return false;
  }
  start(designMode) {
    // key must specified
    const steps = (this.steps || []).filter(i => i);
    const dom = document.createElement('div');
    dom.className = `${this.prefixCls}-stage${designMode ? ' design-mode' : ''}${this.className ?
      ` ${this.className}` : ''}${this.isBlocking ? ' isBlocking' : ''}`;
    document.body.appendChild(dom);
    const overflow = document.body.style.overflowY;
    if (this.isBlocking) {
      document.body.style.overflowY = 'hidden';
    }
    scrollToTop(0);
    this.stop = function stop() {
      ReactDOM.unmountComponentAtNode(dom);
      document.body.removeChild(dom);
      document.body.style.overflowY = overflow;
      this.drop();
      if (typeof this.onComplete === 'function') {
        this.onComplete();
      }
    };
    ReactDOM.render(<UserGuideStage
      steps={steps}
      assistType={this.assistType}
      onAssistClick={this.onAssistClick}
      locale={this.locale}
      prefixCls={this.prefixCls}
      className={this.className}
      done={() => this.stop()}
      designMode={designMode}
    />, dom);
  }
  drop() {
    delete guideMap[this.key];
  }
}

UserGuide.getWithKey = function getWithKey(key, config) {
  if (guideMap[key]) {
    return guideMap[key];
  }
  guideMap[key] = new UserGuide(key, config);
  return guideMap[key];
};

module.exports = UserGuide;


src/AitoUserGuide.js
/**
 * AitoUserGuide Component for uxcore
 * @author buzhou
 *
 * Copyright 2015-2016, Uxcore Team, Alinw.
 * All rights reserved.
 */
const React = require('react');
const ReactDOM = require('react-dom');
const AitoUserGuideStage = require('./AitoUserGuideStage');
const scrollToTop = require('./scrollToTop');

const guideMap = {};

// HOC, 不是一个react component
export default class AitoUserGuide {
  constructor(key, config) {
    this.key = key;
    this.locale = config && config.locale || 'zh-cn';
    this.useImg = config && config.useImage || false;
  }
  addUserGuide(guideProps) {
    const { step = 0 } = guideProps;
    if (!this.steps) {
      this.steps = [];
    }
    this.steps[step] = guideProps || {};
    // if (WrappedComponent instanceof HTMLElement) {
    //   this.steps[step].dom = WrappedComponent;
    //   return WrappedComponent;
    // }
    if (guideProps.type === 'ReactComponent') {
      return (props) => <guideProps.dom
        {...props}
        ref={(comp) => {
          this.steps[step].dom = comp;
        }}
      />;
    }
    return false;
  }
  start(designMode) {
    // key must specified
    const steps = (this.steps || []).filter(i => i);
    const dom = document.createElement('div');
    dom.className = `aito-user-guide-stage${designMode ? ' design-mode' : ''}`;
    document.body.appendChild(dom);
    const overflow = document.body.style.overflowY;
    document.body.style.overflowY = 'hidden';
    scrollToTop(0);
    ReactDOM.render(<AitoUserGuideStage
      steps={steps}
      locale={this.locale}
      done={() => {
        document.body.removeChild(dom);
        document.body.style.overflowY = overflow;
        this.drop();
        if (typeof this.onComplete === 'function') {
          this.onComplete();
        }
      }}
      designMode={designMode}
    />, dom);
  }
  drop() {
    delete guideMap[this.key];
  }
}

AitoUserGuide.getWithKey = function getWithKey(key, config) {
  if (guideMap[key]) {
    return guideMap[key];
  }
  guideMap[key] = new AitoUserGuide(key, config);
  return guideMap[key];
};

module.exports = AitoUserGuide;


demo/AitoUserGuideDemo.js
/**
 * AitoUserGuide Component Demo for uxcore
 * @author buzhou
 *
 * Copyright 2015-2016, Uxcore Team, Alinw.
 * All rights reserved.
 */

const React = require('react');
const AitoUserGuide = require('../src').getWithKey('1');

const Step1 = AitoUserGuide.addUserGuide({
  dom: 'button',
  step: 1,
  title: '我是第一步提示',
  content: '这是一段说明文字，用来补充描述内容',
  contentType: 'TEXT',
  type: 'ReactComponent',
});

AitoUserGuide.addUserGuide({
  step: 2,
  title: '我是第二部的提示，我没有对应的DOM',
  contentType: 'IMAGE',
  content: 'https://timgsa.baidu.com/timg?image&quality=80&size=b9999_10000&sec=1532967222590&di=18571ba9affc82987a08966098a75b26&imgtype=0&src=http%3A%2F%2Fh.hiphotos.baidu.com%2Fimage%2Fpic%2Fitem%2F267f9e2f07082838304837cfb499a9014d08f1a0.jpg',
  type: 'HTMLElementMaker',
  dom() {
    return document.getElementById('justTestIt');
  },
});

AitoUserGuide.addUserGuide({
  step: 3,
  title: '我是第三部的提示，我也没有对应的DOM',
  type: 'Image',
  contentType: 'VIDEO',
  content: 'http://techslides.com/demos/sample-videos/small.mp4',
  poster: 'https://timgsa.baidu.com/timg?image&quality=80&size=b9999_10000&sec=1532968252286&di=702ee1cb712c50b7f41ba87ac6707737&imgtype=0&src=http%3A%2F%2Fp0.qhimgs4.com%2Ft01383c80228884eb05.jpg',
});

class Demo extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
    };
  }

  render() {
    return (
      <div>
        <Step1 onClick={() => AitoUserGuide.start()}>测试的按钮</Step1>
        <div id="justTestIt">我是一个测试holder</div>
      </div>
    );
  }
}

module.exports = Demo;


demo/UserGuideDemo.js
/**
 * UserGuide Component Demo for uxcore
 * @author buzhou
 *
 * Copyright 2015-2016, Uxcore Team, Alinw.
 * All rights reserved.
 */
const React = require('react');
const Button = require('uxcore-button');
const UserGuideFactory = require('../src');
const UserGuide = UserGuideFactory.getWithKey('1', {
  isBlocking: false,
  assistType: 'NO_REMIND',
  onAssistClick(step) {
    console.log(step);
    UserGuide.stop();
  },
});

const Step1 = UserGuide.addUserGuide({
  dom: 'button',
  step: 1,
  title: '我是第一步提示',
  content: '这是一段说明文字，用来补充描述内容',
  iconName: 'shanchu',
  contentType: 'TEXT',
  type: 'ReactComponent',
});

UserGuide.addUserGuide({
  step: 2,
  title: '我是第二步的提示，指定了DIV和图片',
  contentType: 'IMAGE',
  content: 'https://timgsa.baidu.com/timg?image&quality=80&size=b9999_10000&sec=1532967222590&di=18571ba9affc82987a08966098a75b26&imgtype=0&src=http%3A%2F%2Fh.hiphotos.baidu.com%2Fimage%2Fpic%2Fitem%2F267f9e2f07082838304837cfb499a9014d08f1a0.jpg',
  type: 'HTMLElementMaker',
  getDom() {
    return document.getElementById('justTestIt');
  },
});

UserGuide.addUserGuide({
  step: 3,
  title: '我是第三步的提示，我也没有对应的DOM，指定了视频',
  contentType: 'VIDEO',
  content: 'http://techslides.com/demos/sample-videos/small.mp4',
  poster: 'https://timgsa.baidu.com/timg?image&quality=80&size=b9999_10000&sec=1532968252286&di=702ee1cb712c50b7f41ba87ac6707737&imgtype=0&src=http%3A%2F%2Fp0.qhimgs4.com%2Ft01383c80228884eb05.jpg',
});

class Demo extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
    };
  }

  componentDidMount() {
    this.guid2 = UserGuideFactory.getWithKey('2', {
      assistType: 'SKIP',
    });
    this.guid2.addUserGuide({
      step: 1,
      title: '我是第一步提示',
      content: '这是一段说明文字，用来补充描述内容',
      contentType: 'TEXT',
      type: 'HTMLElementMaker',
      getDom() {
        //
        return document.getElementById('btn');
      },
    });
  }

  render() {
    return (
      <div>
        <Step1 onClick={() => UserGuide.start()}>执行非阻塞，带 不在提醒 的引导</Step1>
        <div id="justTestIt">我是一个测试holder</div>
        <Button
          id="btn"
          ref={(ref) => { this.btn = ref; }}
          onClick={() => {
            this.guid2.start();
          }}
        >
          执行阻塞的，带跳过的引导
        </Button>
      </div>
    );
  }
}

module.exports = Demo;
